```
Run appstreamcli validate res/linux/org.mixxx.Mixxx.metainfo.xml
E: org.mixxx.Mixxx:101: release-time-missing date
W: org.mixxx.Mixxx:93: url-not-reachable https://www.mixxx.org/support - Unexpected status code: 301

✘ Validation failed: errors: 1, warnings: 1, pedantic: 1
Error: Process completed with exit code 3.
```
:disappointed:

--- a/.github/workflows/pre-commit.yml
+++ b/.github/workflows/pre-commit.yml
@@ -10,6 +10,9 @@ jobs:
   pre-commit:
     name: Detecting code style issues
     runs-on: ubuntu-latest
+    # The Dockerfile for this container can be found at:
+    # https://github.com/Holzhaus/mixxx-ci-docker
+    container: holzhaus/mixxx-ci:20220805
     steps:
     - name: "Check out repository"
       uses: actions/checkout@v3
@@ -20,18 +23,16 @@ jobs:
         # suffice.
         fetch-depth: 0
 
-    - name: "Set up Python"
-      uses: actions/setup-python@v4
-      with:
-        python-version: '3.10'
-
-    - name: Install clang-format
-      run: sudo apt-get update && sudo apt-get install -y --no-install-recommends clang-format-10
-
-    # Workaround the fact that ubuntu doesn't ship a new enough system
-    # distlib version
-    - name: "Forcefully install newer distlib"
-      run: "python -m pip install --user distlib==0.3.5"
+    - name: "Add GitHub workspace as a safe directory"
+      # Without this, git commands will fail due to mismatching permissions in
+      # the container. See actions/runner#2033 for details.
+      #
+      # The actions/checkout action should already take care of this thanks to
+      # commit actions/checkout@55fd82fc42c0cdd6f1f480dd23f60636a42f6f5c, but
+      # it seems like that's not working properly.
+      run: |
+        git config --global --add safe.directory "${GITHUB_WORKSPACE}"
+        git config --global --list
 
     - name: "Detect code style issues (push)"
       uses: pre-commit/action@v3.0.0
@@ -66,9 +67,6 @@ jobs:
       with:
         name: ${{ env.UPLOAD_PATCH_FILE }}
         path: ${{ env.UPLOAD_PATCH_FILE }}
-
-    - name: "Install appstreamcli"
-      run: sudo apt-get update && sudo apt-get install -y --no-install-recommends appstream
 
     # AppStream metadata has been generated/updated by a pre-commit hook
     - name: "Validate AppStream metadata"

--- a/.github/workflows/pre-commit.yml
+++ b/.github/workflows/pre-commit.yml
@@ -14,7 +14,11 @@ jobs:
     - name: "Check out repository"
       uses: actions/checkout@v3
       with:
-        fetch-depth: 2
+        # Unfortunately we need the whole history and can't use a shallow clone
+        # because the Appstream Metadata hook parses the history to find the
+        # latest changelog modification date. Otherwise, `fetch-depth: 2` would
+        # suffice.
+        fetch-depth: 0
 
     - name: "Set up Python"
       uses: actions/setup-python@v4

--- a/res/linux/org.mixxx.Mixxx.metainfo.xml
+++ b/res/linux/org.mixxx.Mixxx.metainfo.xml
@@ -89,8 +89,8 @@
   </screenshots>
   <url type="homepage">https://mixxx.org</url>
   <url type="bugtracker">https://bugs.launchpad.net/mixxx</url>
-  <url type="donation">https://mixxx.org/donate</url>
-  <url type="help">https://www.mixxx.org/support</url>
+  <url type="donation">https://mixxx.org/donate/</url>
+  <url type="help">https://mixxx.org/support/</url>
   <url type="translate">https://explore.transifex.com/mixxx-dj-software/</url>
   <url type="contact">https://mixxx.zulipchat.com</url>
   <!--

--- a/res/linux/org.mixxx.Mixxx.metainfo.xml
+++ b/res/linux/org.mixxx.Mixxx.metainfo.xml
@@ -98,7 +98,7 @@
   Do not edit it manually.
   -->
   <releases>
-    <release version="2.3.4" type="development">
+    <release version="2.3.4" type="development" date="2022-08-04" timestamp="1659591032">
       <description>
  <ul>
   <li>

--- a/tools/update_metainfo.py
+++ b/tools/update_metainfo.py
@@ -1,6 +1,7 @@
 #!/usr/bin/env python3
 import datetime
 import os
+import subprocess
 import re
 
 import markdown
@@ -8,7 +9,7 @@ import bs4
 from lxml import etree
 
 
-def parse_changelog(content):
+def parse_changelog(content, development_release_date):
     for section in re.split("^## ", content.strip(), flags=re.MULTILINE)[1:]:
         version, sep, description_md = section.partition("\n")
         matchobj = re.match(
@@ -27,11 +28,15 @@ def parse_changelog(content):
                 matchobj.group("date"), " (%Y-%m-%d)"
             ).replace(tzinfo=datetime.timezone.utc)
         except ValueError:
+            # Every release must have an associated timestamp, even development
+            # releases. Otherwise, `appstreamcli validate` will complain.
+            release_date = development_release_date
             attrib["type"] = "development"
         else:
             attrib["type"] = "stable"
-            attrib["date"] = release_date.strftime("%Y-%m-%d")
-            attrib["timestamp"] = "{:.0f}".format(release_date.timestamp())
+
+        attrib["date"] = release_date.strftime("%Y-%m-%d")
+        attrib["timestamp"] = "{:.0f}".format(release_date.timestamp())
 
         soup = bs4.BeautifulSoup(
             markdown.markdown(description_md), "html.parser"
@@ -74,8 +79,29 @@ def main(argv=None):
     root = tree.getroot()
     releases = root.find("releases")
     releases.clear()
-    with open(os.path.join(rootpath, "CHANGELOG.md"), mode="r") as fp:
-        for release in parse_changelog(fp.read()):
+    changelog_path = os.path.join(rootpath, "CHANGELOG.md")
+
+    # We need to define a date for development releases. We can't use
+    # `datetime.datetime.now()` or something like that, because that would
+    # involve updating the metainfo.xml during every single commit. The same
+    # goes for using the latest commit date.
+    #
+    # As a compromise, we're using the date of the commit that last updated the
+    # changelog here. The only feasible alternative is to remove development
+    # releases completely, or remove the appstream metadata from the repository
+    # and only generate it on demand when an actual package is built.
+    last_changelog_change_date = datetime.datetime.fromisoformat(
+        subprocess.check_output(
+            ("git", "log", "-1", "--format=%cI", "--", changelog_path)
+        )
+        .decode()
+        .strip()
+    )
+
+    with open(changelog_path, mode="r") as fp:
+        for release in parse_changelog(
+            fp.read(), development_release_date=last_changelog_change_date
+        ):
             releases.append(release)
     tree.write(
         metainfo_path,

